### ISSUE TITLE
Run QuadTreeAttention on CPU

### ISSUE THREAD

**Issue by hmarichalds** [Tangshitao/QuadTreeAttention#12]
Hi,

Thank for your work. We are getting very interesting feature matching results in our tests. 

Now we would like the algorithm running on cpu. We made some changes (over torch) in order to run on cpu, but we are getting below  error. Is there a score_computation function implemented to run in cpu?

 File "../../QuadTreeAttention/QuadTreeAttention/QuadtreeAttention/functions/quadtree_attention.py", line 11, in forward
    x = score_computation_cuda.score_forward(query, key, index)

**Comment by Tangshitao** [comment]
I did not implement a cpu version. I am thinking to implement QuadtreeAttention with torch.geometry (https://pytorch-geometric.readthedocs.io/en/latest/), but I am too busy recently to implement it.

**Comment by hmarichalds** [comment]
Thanks for your anwser! From what I  see, in order to run in cpu I need to migrate the class ScoreComputational and value_agregation (functions/quadtree_attention.py) to full python(torch,numpy) because both of them have the forward and backward methods implemented in cuda, am I right?

**Comment by Tangshitao** [comment]
Yes, you are right. I have a pytorch version of QuadtreeAttention, but the problem is that it's too slow and memory inefficient.  That's why I want to reimplement it with torch.geometry. I can share it with you if you don't care about the inefficiency.

**Comment by hmarichalds** [comment]
Yes, please. I would like to test it. Thanks very much!

**Comment by Tangshitao** [comment]
Here is the code. If you can run the codes without any problem, I would appreciate it if you can make a merge request.
`class QTAttAPytorch(nn.Module):
    def __init__(self, nhead, dim, scale, use_dropout=False, attention_dropout=0.1, topk=1):
        super().__init__()
        self.use_dropout = use_dropout
        self.topk = topk
        self.nhead = nhead
        self.dim = dim

    def forward(self, queries, keys, values, q_mask=None, kv_mask=None):
        """Quadtree attention
        Args:
            query: [N, L, H, D]
            keys: [N, S, H, D]
            values: [N, S, H, D]
            q_mask: [N, L]
            kv_mask: [N, S]
        Returns:
            queried_values: (N, L, H, D)
        """

        bs = queries[0].shape[0]
        # Compute the unnormalized attention and apply the masks
        message = 0
        for i, (query, key, value) in enumerate(zip(reversed(queries), reversed(keys), reversed(values))):
            bs, c, h, w = key.shape
            key = rearrange(key, "b c h w -> b (h w) c").view(bs, -1, self.nhead, self.dim)  # [N, S, H, D]
            value = rearrange(value, "b c h w -> b (h w) c").view(bs, -1, self.nhead, self.dim)  # [N, S, H, D]
            value_old = value
            if i == 0:
                query = rearrange(query, "b c h w -> b (h w) c").view(bs, -1, self.nhead, self.dim)
                QK = torch.einsum("nlhd,nshd->nlsh", query, key)
            else:
                query = query.view(bs, c, h // 2, 2, w // 2, 2)
                query = rearrange(query, "b c h t1 w t2-> b (h w) (t1 t2) c ").view(bs, -1, 4, self.nhead, self.dim)
                topk_pos = topk_pos * 2
                key_gather = []
                value_gather = []
                idx_gather = []

                for x in [0, 1]:
                    for y in [0, 1]:
                        idx = (topk_pos[0] + x) * w + topk_pos[1] + y  # convert to index

                        idx_gather.append(idx)
                        idx = idx.view(bs, -1, self.nhead, 1).repeat(1, 1, 1, self.dim)  # [N, L, K, H, D]

                        k = torch.gather(key, index=idx, dim=1).view(bs, -1, self.topk, self.nhead, self.dim)

                        v = torch.gather(value, index=idx, dim=1).view(bs, -1, self.topk, self.nhead, self.dim)
                        key_gather.append(k)
                        value_gather.append(v)
                idx = torch.stack(idx_gather, dim=3)  # [N, L, K, 4, H, D]

                # query: [b, N, 4, H, D]
                # key: [b, 4N, H, D]
                # idx: [b, N, 4K, H, D]
                # QK: [b, N, 4, K, 4, H]

                key_gather = torch.stack(key_gather, dim=3)  # [N, L, K, 4, H, D]
                value = torch.stack(value_gather, dim=3)  # [N, L, K, 4, H, D]

                QK = torch.einsum("nlwhd,nlkfhd->nlwkfh", query, key_gather)

            softmax_temp = 1.0 / self.dim ** 0.5  # sqrt(D)
            A = torch.softmax(softmax_temp * QK, dim=-2)  # [N, L//scale**i, K, 4, H]

            if i != 0:
                # A: [b, N, 4, K, 4, H]
                # topk_score: [b, N, 1, K, 1, H]
                # return: A, [b, N, 4, K*4, H]
                A = (A * topk_score.unsqueeze(-2).unsqueeze(2)).reshape(bs, -1, 4, self.topk * 4, self.nhead)
                # value: [b, N, K, 4, H, D]
                value = value.reshape(bs, -1, self.topk * 4, self.nhead, self.dim)  # [N, L, K*4, H, D]

            topk_score, topk_idx = torch.topk(A, dim=-2, k=self.topk, largest=True)  # [N, L, 4, K, H]

            mask = torch.ones_like(A)
            if i != len(keys) - 1:
                mask = mask.scatter(dim=-2, index=topk_idx, src=torch.zeros_like(topk_idx).float())
            if i == 0:
                message += torch.einsum("nlsh,nshd->nlhd", A * mask, value)
                # [b, N, H, D]
            else:
                # A: [b, N, 4, 4K, H]
                # value: [b, N, 4K, H, D]
                # message: [b, N, 4, H, D]
                new_message = torch.einsum("nlwkh,nlkhd->nlwhd", A * mask, value)

                idx = idx.view(bs, -1, 1, self.topk * 4, self.nhead).repeat(1, 1, 4, 1, 1)  # [N, L,4, K*4, H]
                # A: [b, N, 4, K*4, H]
                # value_old: [b, 4N, H, D]
                # idx: [b, N, 4, 4K, H]
            
                message = message.unsqueeze(2) + new_message
                message = message.view(bs, h // 2, w // 2, 2, 2, self.nhead, self.dim)
                message = rearrange(message, "b h w t1 t2 nh d -> b (h t1 w t2) nh d")  # reshape

                topk_idx = torch.gather(idx, index=topk_idx, dim=-2)

                topk_idx = topk_idx.view(bs, h // 2, w // 2, 2, 2, self.topk, self.nhead)

                topk_idx = rearrange(topk_idx, "b h w t1 t2 k nh -> b (h t1 w t2) k nh")  # reshape back

                topk_score = topk_score.view(bs, h // 2, w // 2, 2, 2, self.topk, self.nhead)
                topk_score = rearrange(topk_score, "b h w t1 t2 k nh -> b (h t1 w t2) k nh")  # reshape back

            topk_pos = torch.stack([topk_idx // w, topk_idx % w])  # convert to coordinate

        return message`

**Comment by hmarichalds** [comment]
Thanks, I will test it.

**Comment by xmba15** [comment]
@hmarichalds @Tangshitao sorry for asking before testing anything. But can you please confirm if the code above works, regardless of computing speed?

**Comment by Tangshitao** [comment]
Yes, it will work

**Comment by xmba15** [comment]
@Tangshitao Thanks for the confirmation.

**Comment by ZhiWei-Liu123** [comment]
@hmarichalds  @xmba15 Could you share me with the weights file? When I run pytorch version of QuadtreeAttention using the original weight file, it will occur the following problem.
![image](https://github.com/Tangshitao/QuadTreeAttention/assets/133549543/35cf249f-7a1b-423a-a986-398be3f2613a)
